File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_MonsterBait.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.map.Area;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.monster.MonsterBehavior;
import com.broll.gainea.server.core.utils.LocationUtils;
import com.broll.gainea.server.core.utils.UnitControl;

import java.util.List;
import java.util.stream.Collectors;

public class C_MonsterBait extends Card {
    public C_MonsterBait() {
        super(69, "Köderstein", "Wählt ein neutrales Gebiet, alle benachbarten Monster bewegen sich dorthin und werden sesshaft.");
    }

    @Override
    public boolean isPlayable() {
        return true;
    }

    @Override
    protected void play() {
        List<Area> areas = game.getMap().getAllAreas().stream().filter(LocationUtils::emptyOrWildMonster).collect(Collectors.toList());
        Location location = selectHandler.selectLocation(owner, "Wähle einen Zielort", areas);
        location.getConnectedLocations().stream().flatMap(it -> LocationUtils.getMonsters(it).stream())
                .forEach(monster -> {
                    monster.setBehavior(MonsterBehavior.RESIDENT);
                    UnitControl.move(game, monster, location);
                });
    }
}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_MonsterBait.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.cards.EffectType.DISRUPTION
import com.broll.gainea.server.core.objects.monster.MonsterBehavior
import com.broll.gainea.server.core.utils.UnitControl.move
import com.broll.gainea.server.core.utils.emptyOrWildMonster
import com.broll.gainea.server.core.utils.getMonsters

class C_MonsterBait : Card(
    69,
    DISRUPTION,
    "Köderstein",
    "Wählt ein neutrales Gebiet, alle benachbarten Monster bewegen sich dorthin und werden sesshaft."
) {
    override val isPlayable: Boolean
        get() = true

    override fun play() {
        val areas = game.map.allAreas.filter { it.emptyOrWildMonster() }
        val location = selectHandler.selectLocation(owner, "Wähle einen Zielort", areas)
        location.connectedLocations.flatMap { it.getMonsters() }
            .forEach { monster ->
                monster.behavior = MonsterBehavior.RESIDENT
                game.move(monster, location)
            }
    }
}
